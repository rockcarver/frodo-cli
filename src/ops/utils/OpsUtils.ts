import { FrodoError } from '@rockcarver/frodo-lib';

import { printError } from '../../utils/Console';

/**
 * Deep clone object
 * @param {any} obj object to deep clone
 * @returns {any} new object cloned from obj
 */
export function cloneDeep(obj: any): any {
  return JSON.parse(JSON.stringify(obj));
}

/**
 * Result callback that handles errors when they occur by printing them and setting the process exit code
 * @param error The error to handle
 */
export function errorHandler(error?: FrodoError) {
  if (!error) return;
  printError(error);
  process.exitCode = 1;
}
