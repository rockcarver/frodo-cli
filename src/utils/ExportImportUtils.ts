import fs from 'fs';
import slugify from 'slugify';
import { ExportImportUtils, state } from '@rockcarver/frodo-lib';
import { printMessage } from './Console';

const { getMetadata } = ExportImportUtils;

/**
 * Get a typed filename. E.g. "my-script.script.json"
 *
 * @param name The name of the file
 * @param type The type of the file, e.g. script, idp, etc.
 * @param suffix The suffix of the file, e.g. json, xml, etc. Defaults to json.
 * @returns The typed filename
 */
export function getTypedFilename(
  name: string,
  type: string,
  suffix = 'json'
): string {
  const slug = slugify(name.replace(/^http(s?):\/\//, ''));
  return `${slug}.${type}.${suffix}`;
}

/**
 * Save JSON object to file
 *
 * @param data data object
 * @param filename file name
 */
export function saveJsonToFile(data: any, filename: string) {
  const exportData = data;
  if (!exportData.meta) {
    exportData.meta = getMetadata();
  }
  fs.writeFile(filename, JSON.stringify(exportData, null, 2), (err) => {
    if (err) {
      return printMessage(`ERROR - can't save ${filename}`, 'error');
    }
    return '';
  });
}

export function saveToFile(type, data, identifier, filename) {
  const exportData = {};
  exportData['meta'] = getMetadata();
  exportData[type] = {};

  if (Array.isArray(data)) {
    data.forEach((element) => {
      exportData[type][element[identifier]] = element;
    });
  } else {
    exportData[type][data[identifier]] = data;
  }
  fs.writeFile(filename, JSON.stringify(exportData, null, 2), (err) => {
    if (err) {
      return printMessage(`ERROR - can't save ${type} to file`, 'error');
    }
    return '';
  });
}

/**
 * Save text data to file
 * @param data text data
 * @param filename file name
 * @return true if successful, false otherwise
 */
export function saveTextToFile(data: string, filename: string): boolean {
  try {
    fs.writeFileSync(filename, data);
    return true;
  } catch (error) {
    printMessage(`ERROR - can't save ${filename}`, 'error');
    return false;
  }
}

/*
 * Output str in title case
 *
 * e.g.: 'ALL UPPERCASE AND all lowercase' = 'All Uppercase And All Lowercase'
 */
export function titleCase(input) {
  const str = input.toString();
  const splitStr = str.toLowerCase().split(' ');
  for (let i = 0; i < splitStr.length; i += 1) {
    splitStr[i] = splitStr[i].charAt(0).toUpperCase() + splitStr[i].slice(1);
  }
  return splitStr.join(' ');
}

export function getRealmString() {
  const realm = state.getRealm();
  return realm
    .split('/')
    .reduce((result, item) => `${result}${titleCase(item)}`, '');
}
