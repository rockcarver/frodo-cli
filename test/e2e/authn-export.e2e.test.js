/**
 * Follow this process to write e2e tests for the CLI project:
 *
 * 1. Test if all the necessary mocks for your tests already exist.
 *    In mock mode, run the command you want to test with the same arguments
 *    and parameters exactly as you want to test it, for example:
 *
 *    $ FRODO_MOCK=1 frodo conn save https://openam-frodo-dev.forgeblocks.com/am volker.scheuber@forgerock.com Sup3rS3cr3t!
 *
 *    If your command completes without errors and with the expected results,
 *    all the required mocks already exist and you are good to write your
 *    test and skip to step #4.
 *
 *    If, however, your command fails and you see errors like the one below,
 *    you know you need to record the mock responses first:
 *
 *    [Polly] [adapter:node-http] Recording for the following request is not found and `recordIfMissing` is `false`.
 *
 * 2. Record mock responses for your exact command.
 *    In mock record mode, run the command you want to test with the same arguments
 *    and parameters exactly as you want to test it, for example:
 *
 *    $ FRODO_MOCK=record frodo conn save https://openam-frodo-dev.forgeblocks.com/am volker.scheuber@forgerock.com Sup3rS3cr3t!
 *
 *    Wait until you see all the Polly instances (mock recording adapters) have
 *    shutdown before you try to run step #1 again.
 *    Messages like these indicate mock recording adapters shutting down:
 *
 *    Polly instance 'conn/4' stopping in 3s...
 *    Polly instance 'conn/4' stopping in 2s...
 *    Polly instance 'conn/save/3' stopping in 3s...
 *    Polly instance 'conn/4' stopping in 1s...
 *    Polly instance 'conn/save/3' stopping in 2s...
 *    Polly instance 'conn/4' stopped.
 *    Polly instance 'conn/save/3' stopping in 1s...
 *    Polly instance 'conn/save/3' stopped.
 *
 * 3. Validate your freshly recorded mock responses are complete and working.
 *    Re-run the exact command you want to test in mock mode (see step #1).
 *
 * 4. Write your test.
 *    Make sure to use the exact command including number of arguments and params.
 *
 * 5. Commit both your test and your new recordings to the repository.
 *    Your tests are likely going to reside outside the frodo-lib project but
 *    the recordings must be committed to the frodo-lib project.
 */

/*
// Cloud
FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo authn export -ND authnExportDir1
FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo authn export -f authnExportTest.json
FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo authn export --no-metadata --file authnExportTest.json --directory  authnExportDir2
// Classic
FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=http://openam-frodo-dev.classic.com:8080/am frodo authn export -g -m classic
FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=http://openam-frodo-dev.classic.com:8080/am frodo authn export --global --no-metadata --directory authnExportDir3 --type classic
*/
import { getEnv, testExport } from './utils/TestUtils';
import { connection as c, classic_connection as cc } from './utils/TestConfig';

process.env['FRODO_MOCK'] = '1';
process.env['FRODO_CONNECTION_PROFILES_PATH'] =
    './test/e2e/env/Connections.json';
const env = getEnv(c);
const classicEnv = getEnv(cc);

const type = 'authentication.settings';

describe('frodo authn export', () => {
    test('"frodo authn export -ND authnExportDir1": should export authentication settings to a single file in the "authnExportDir1" directory', async () => {
        const exportFile = "alphaRealm.authentication.settings.json";
        const exportDirectory = "authnExportDir1";
        const CMD = `frodo authn export -ND ${exportDirectory}`;
        await testExport(CMD, env, type, exportFile, exportDirectory, false);
    });

    test('"frodo authn export -f authnExportTest.json": should export authentication settings to a file named "authnExportTest.json"', async () => {
        const exportFile = "authnExportTest.json";
        const CMD = `frodo authn export -f ${exportFile}`;
        await testExport(CMD, env, type, exportFile);
    });

    test('"frodo authn export --no-metadata --file authnExportTest.json --directory  authnExportDir2": should export authentication settings to a file named "authnExportTest.json" in the "authnExportDir2" directory', async () => {
        const exportFile = "authnExportTest.json";
        const exportDirectory = "authnExportDir2";
        const CMD = `frodo authn export --no-metadata --file ${exportFile} --directory ${exportDirectory}`;
        await testExport(CMD, env, type, exportFile, exportDirectory, false);
    });

    test('"frodo authn export -g -m classic": should export global authentication settings to a file', async () => {
        const exportFile = "global.authentication.settings.json";
        const CMD = `frodo authn export -g -m classic`;
        await testExport(CMD, classicEnv, type, exportFile);
    });

    test('"frodo authn export --global --no-metadata --directory  authnExportDir3 --type classic": should export global authentication settings to a file in the "authnExportDir3" directory', async () => {
        const exportFile = "global.authentication.settings.json";
        const exportDirectory = "authnExportDir3";
        const CMD = `frodo authn export --global --no-metadata --directory ${exportDirectory} --type classic`;
        await testExport(CMD, classicEnv, type, exportFile, exportDirectory, false);
    });
});
