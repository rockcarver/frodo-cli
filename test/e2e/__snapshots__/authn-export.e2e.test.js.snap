// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frodo authn export "frodo authn export --global --no-metadata --directory  authnExportDir3 --type classic": should export global authentication settings to a file in the "authnExportDir3" directory 1`] = `0`;

exports[`frodo authn export "frodo authn export --global --no-metadata --directory  authnExportDir3 --type classic": should export global authentication settings to a file in the "authnExportDir3" directory 2`] = `""`;

exports[`frodo authn export "frodo authn export --global --no-metadata --directory  authnExportDir3 --type classic": should export global authentication settings to a file in the "authnExportDir3" directory: authnExportDir3/global.authentication.settings.json 1`] = `
{
  "authentication": {
    "_id": "",
    "_type": {
      "_id": "EMPTY",
      "collection": false,
      "name": "Core",
    },
    "authenticators": [
      "com.sun.identity.authentication.modules.ad.AD",
      "org.forgerock.openam.authentication.modules.saml2.SAML2",
      "org.forgerock.openam.authentication.modules.social.SocialAuthInstagram",
      "org.forgerock.openam.authentication.modules.oath.OATH",
      "org.forgerock.openam.authentication.modules.social.SocialAuthVK",
      "com.sun.identity.authentication.modules.membership.Membership",
      "com.sun.identity.authentication.modules.windowsdesktopsso.WindowsDesktopSSO",
      "org.forgerock.openam.authentication.modules.deviceprint.DeviceIdSave",
      "com.sun.identity.authentication.modules.federation.Federation",
      "org.forgerock.openam.authentication.modules.deviceprint.DeviceIdMatch",
      "com.sun.identity.authentication.modules.jdbc.JDBC",
      "com.sun.identity.authentication.modules.radius.RADIUS",
      "com.sun.identity.authentication.modules.anonymous.Anonymous",
      "com.sun.identity.authentication.modules.cert.Cert",
      "org.forgerock.openam.authentication.modules.push.registration.AuthenticatorPushRegistration",
      "com.sun.identity.authentication.modules.httpbasic.HTTPBasic",
      "org.forgerock.openam.authentication.modules.oidc.OpenIdConnect",
      "com.sun.identity.authentication.modules.sae.SAE",
      "org.forgerock.openam.authentication.modules.social.SocialAuthWeChat",
      "org.forgerock.openam.authentication.modules.persistentcookie.PersistentCookie",
      "org.forgerock.openam.authentication.modules.social.SocialAuthTwitter",
      "com.sun.identity.authentication.modules.ldap.LDAP",
      "org.forgerock.openam.authentication.modules.push.AuthenticatorPush",
      "org.forgerock.openam.authentication.modules.oauth2.OAuth",
      "com.sun.identity.authentication.modules.nt.NT",
      "org.forgerock.openam.authentication.modules.social.SocialAuthWeChatMobile",
      "org.forgerock.openam.authentication.modules.jwtpop.JwtProofOfPossession",
      "com.sun.identity.authentication.modules.application.Application",
      "org.forgerock.openam.authentication.modules.scripted.Scripted",
      "org.forgerock.openam.authentication.modules.social.SocialAuthOAuth2",
      "com.sun.identity.authentication.modules.hotp.HOTP",
      "org.forgerock.openam.authentication.modules.adaptive.Adaptive",
      "org.forgerock.openam.authentication.modules.accountactivecheck.AccountActiveCheck",
      "org.forgerock.openam.authentication.modules.social.SocialAuthOpenID",
      "com.sun.identity.authentication.modules.msisdn.MSISDN",
      "org.forgerock.openam.authentication.modules.fr.oath.AuthenticatorOATH",
      "com.sun.identity.authentication.modules.datastore.DataStore",
      "com.sun.identity.authentication.modules.securid.SecurID",
      "org.forgerock.openam.authentication.modules.amster.Amster",
    ],
    "defaults": {
      "accountlockout": {
        "lockoutDuration": 0,
        "lockoutDurationMultiplier": 1,
        "lockoutWarnUserCount": 0,
        "loginFailureCount": 5,
        "loginFailureDuration": 300,
        "loginFailureLockoutMode": false,
        "storeInvalidAttemptsInDataStore": true,
      },
      "core": {
        "adminAuthModule": "[Empty]",
        "orgConfig": "[Empty]",
      },
      "general": {
        "defaultAuthLevel": 0,
        "identityType": [
          "agent",
          "user",
        ],
        "locale": "en_US",
        "statelessSessionsEnabled": false,
        "twoFactorRequired": false,
        "userStatusCallbackPlugins": [],
      },
      "postauthprocess": {
        "loginFailureUrl": [],
        "loginPostProcessClass": [],
        "loginSuccessUrl": [
          "/am/console",
        ],
        "userAttributeSessionMapping": [],
        "usernameGeneratorClass": "com.sun.identity.authentication.spi.DefaultUserIDGenerator",
        "usernameGeneratorEnabled": true,
      },
      "security": {
        "addClearSiteDataHeader": true,
        "moduleBasedAuthEnabled": true,
        "sharedSecret": null,
        "zeroPageLoginAllowedWithoutReferrer": true,
        "zeroPageLoginEnabled": false,
        "zeroPageLoginReferrerWhiteList": [],
      },
      "trees": {
        "authenticationSessionsMaxDuration": 5,
        "authenticationSessionsStateManagement": "JWT",
        "authenticationSessionsWhitelist": false,
        "authenticationTreeCookieHttpOnly": true,
        "suspendedAuthenticationTimeout": 5,
      },
      "userprofile": {
        "aliasAttributeName": [],
        "defaultRole": [],
        "dynamicProfileCreation": "false",
      },
    },
    "keepPostProcessInstances": false,
    "ldapConnectionPoolDefaultSize": "1:10",
    "ldapConnectionPoolSize": [],
    "remoteAuthSecurityEnabled": false,
  },
}
`;

exports[`frodo authn export "frodo authn export --no-metadata --file authnExportTest.json --directory  authnExportDir2": should export authentication settings to a file named "authnExportTest.json" in the "authnExportDir2" directory 1`] = `0`;

exports[`frodo authn export "frodo authn export --no-metadata --file authnExportTest.json --directory  authnExportDir2": should export authentication settings to a file named "authnExportTest.json" in the "authnExportDir2" directory 2`] = `""`;

exports[`frodo authn export "frodo authn export --no-metadata --file authnExportTest.json --directory  authnExportDir2": should export authentication settings to a file named "authnExportTest.json" in the "authnExportDir2" directory: authnExportDir2/authnExportTest.json 1`] = `
{
  "authentication": {
    "_id": "",
    "_type": {
      "_id": "EMPTY",
      "collection": false,
      "name": "Core",
    },
    "accountlockout": {
      "lockoutDuration": 0,
      "lockoutDurationMultiplier": 1,
      "lockoutWarnUserCount": 0,
      "loginFailureCount": 5,
      "loginFailureDuration": 300,
      "loginFailureLockoutMode": false,
      "storeInvalidAttemptsInDataStore": true,
    },
    "core": {
      "adminAuthModule": "Login",
      "orgConfig": "Login",
    },
    "general": {
      "defaultAuthLevel": 0,
      "identityType": [
        "agent",
        "user",
      ],
      "locale": "en_US",
      "statelessSessionsEnabled": false,
      "twoFactorRequired": false,
      "userStatusCallbackPlugins": [],
    },
    "postauthprocess": {
      "loginFailureUrl": [],
      "loginPostProcessClass": [],
      "loginSuccessUrl": [
        "/enduser/?realm=/alpha",
      ],
      "userAttributeSessionMapping": [],
      "usernameGeneratorClass": "com.sun.identity.authentication.spi.DefaultUserIDGenerator",
      "usernameGeneratorEnabled": true,
    },
    "security": {
      "addClearSiteDataHeader": true,
      "keyAlias": "test",
      "moduleBasedAuthEnabled": false,
      "sharedSecret": {
        "$string": "&{am.authentication.shared.secret}",
      },
      "zeroPageLoginAllowedWithoutReferrer": true,
      "zeroPageLoginEnabled": false,
      "zeroPageLoginReferrerWhiteList": [],
    },
    "trees": {
      "authenticationSessionsMaxDuration": 5,
      "authenticationSessionsStateManagement": "JWT",
      "authenticationSessionsWhitelist": false,
      "authenticationTreeCookieHttpOnly": true,
      "suspendedAuthenticationTimeout": 1440,
    },
    "userprofile": {
      "aliasAttributeName": [
        "uid",
      ],
      "defaultRole": [],
      "dynamicProfileCreation": "false",
    },
  },
}
`;

exports[`frodo authn export "frodo authn export -ND authnExportDir1": should export authentication settings to a single file in the "authnExportDir1" directory 1`] = `0`;

exports[`frodo authn export "frodo authn export -ND authnExportDir1": should export authentication settings to a single file in the "authnExportDir1" directory 2`] = `""`;

exports[`frodo authn export "frodo authn export -ND authnExportDir1": should export authentication settings to a single file in the "authnExportDir1" directory: authnExportDir1/alphaRealm.authentication.settings.json 1`] = `
{
  "authentication": {
    "_id": "",
    "_type": {
      "_id": "EMPTY",
      "collection": false,
      "name": "Core",
    },
    "accountlockout": {
      "lockoutDuration": 0,
      "lockoutDurationMultiplier": 1,
      "lockoutWarnUserCount": 0,
      "loginFailureCount": 5,
      "loginFailureDuration": 300,
      "loginFailureLockoutMode": false,
      "storeInvalidAttemptsInDataStore": true,
    },
    "core": {
      "adminAuthModule": "Login",
      "orgConfig": "Login",
    },
    "general": {
      "defaultAuthLevel": 0,
      "identityType": [
        "agent",
        "user",
      ],
      "locale": "en_US",
      "statelessSessionsEnabled": false,
      "twoFactorRequired": false,
      "userStatusCallbackPlugins": [],
    },
    "postauthprocess": {
      "loginFailureUrl": [],
      "loginPostProcessClass": [],
      "loginSuccessUrl": [
        "/enduser/?realm=/alpha",
      ],
      "userAttributeSessionMapping": [],
      "usernameGeneratorClass": "com.sun.identity.authentication.spi.DefaultUserIDGenerator",
      "usernameGeneratorEnabled": true,
    },
    "security": {
      "addClearSiteDataHeader": true,
      "keyAlias": "test",
      "moduleBasedAuthEnabled": false,
      "sharedSecret": {
        "$string": "&{am.authentication.shared.secret}",
      },
      "zeroPageLoginAllowedWithoutReferrer": true,
      "zeroPageLoginEnabled": false,
      "zeroPageLoginReferrerWhiteList": [],
    },
    "trees": {
      "authenticationSessionsMaxDuration": 5,
      "authenticationSessionsStateManagement": "JWT",
      "authenticationSessionsWhitelist": false,
      "authenticationTreeCookieHttpOnly": true,
      "suspendedAuthenticationTimeout": 1440,
    },
    "userprofile": {
      "aliasAttributeName": [
        "uid",
      ],
      "defaultRole": [],
      "dynamicProfileCreation": "false",
    },
  },
}
`;

exports[`frodo authn export "frodo authn export -f authnExportTest.json": should export authentication settings to a file named "authnExportTest.json" 1`] = `0`;

exports[`frodo authn export "frodo authn export -f authnExportTest.json": should export authentication settings to a file named "authnExportTest.json" 2`] = `""`;

exports[`frodo authn export "frodo authn export -f authnExportTest.json": should export authentication settings to a file named "authnExportTest.json": authnExportTest.json 1`] = `
{
  "authentication": {
    "_id": "",
    "_type": {
      "_id": "EMPTY",
      "collection": false,
      "name": "Core",
    },
    "accountlockout": {
      "lockoutDuration": 0,
      "lockoutDurationMultiplier": 1,
      "lockoutWarnUserCount": 0,
      "loginFailureCount": 5,
      "loginFailureDuration": 300,
      "loginFailureLockoutMode": false,
      "storeInvalidAttemptsInDataStore": true,
    },
    "core": {
      "adminAuthModule": "Login",
      "orgConfig": "Login",
    },
    "general": {
      "defaultAuthLevel": 0,
      "identityType": [
        "agent",
        "user",
      ],
      "locale": "en_US",
      "statelessSessionsEnabled": false,
      "twoFactorRequired": false,
      "userStatusCallbackPlugins": [],
    },
    "postauthprocess": {
      "loginFailureUrl": [],
      "loginPostProcessClass": [],
      "loginSuccessUrl": [
        "/enduser/?realm=/alpha",
      ],
      "userAttributeSessionMapping": [],
      "usernameGeneratorClass": "com.sun.identity.authentication.spi.DefaultUserIDGenerator",
      "usernameGeneratorEnabled": true,
    },
    "security": {
      "addClearSiteDataHeader": true,
      "keyAlias": "test",
      "moduleBasedAuthEnabled": false,
      "sharedSecret": {
        "$string": "&{am.authentication.shared.secret}",
      },
      "zeroPageLoginAllowedWithoutReferrer": true,
      "zeroPageLoginEnabled": false,
      "zeroPageLoginReferrerWhiteList": [],
    },
    "trees": {
      "authenticationSessionsMaxDuration": 5,
      "authenticationSessionsStateManagement": "JWT",
      "authenticationSessionsWhitelist": false,
      "authenticationTreeCookieHttpOnly": true,
      "suspendedAuthenticationTimeout": 1440,
    },
    "userprofile": {
      "aliasAttributeName": [
        "uid",
      ],
      "defaultRole": [],
      "dynamicProfileCreation": "false",
    },
  },
  "meta": Any<Object>,
}
`;

exports[`frodo authn export "frodo authn export -g -m classic": should export global authentication settings to a file 1`] = `0`;

exports[`frodo authn export "frodo authn export -g -m classic": should export global authentication settings to a file 2`] = `""`;

exports[`frodo authn export "frodo authn export -g -m classic": should export global authentication settings to a file: global.authentication.settings.json 1`] = `
{
  "authentication": {
    "_id": "",
    "_type": {
      "_id": "EMPTY",
      "collection": false,
      "name": "Core",
    },
    "authenticators": [
      "com.sun.identity.authentication.modules.ad.AD",
      "org.forgerock.openam.authentication.modules.saml2.SAML2",
      "org.forgerock.openam.authentication.modules.social.SocialAuthInstagram",
      "org.forgerock.openam.authentication.modules.oath.OATH",
      "org.forgerock.openam.authentication.modules.social.SocialAuthVK",
      "com.sun.identity.authentication.modules.membership.Membership",
      "com.sun.identity.authentication.modules.windowsdesktopsso.WindowsDesktopSSO",
      "org.forgerock.openam.authentication.modules.deviceprint.DeviceIdSave",
      "com.sun.identity.authentication.modules.federation.Federation",
      "org.forgerock.openam.authentication.modules.deviceprint.DeviceIdMatch",
      "com.sun.identity.authentication.modules.jdbc.JDBC",
      "com.sun.identity.authentication.modules.radius.RADIUS",
      "com.sun.identity.authentication.modules.anonymous.Anonymous",
      "com.sun.identity.authentication.modules.cert.Cert",
      "org.forgerock.openam.authentication.modules.push.registration.AuthenticatorPushRegistration",
      "com.sun.identity.authentication.modules.httpbasic.HTTPBasic",
      "org.forgerock.openam.authentication.modules.oidc.OpenIdConnect",
      "com.sun.identity.authentication.modules.sae.SAE",
      "org.forgerock.openam.authentication.modules.social.SocialAuthWeChat",
      "org.forgerock.openam.authentication.modules.persistentcookie.PersistentCookie",
      "org.forgerock.openam.authentication.modules.social.SocialAuthTwitter",
      "com.sun.identity.authentication.modules.ldap.LDAP",
      "org.forgerock.openam.authentication.modules.push.AuthenticatorPush",
      "org.forgerock.openam.authentication.modules.oauth2.OAuth",
      "com.sun.identity.authentication.modules.nt.NT",
      "org.forgerock.openam.authentication.modules.social.SocialAuthWeChatMobile",
      "org.forgerock.openam.authentication.modules.jwtpop.JwtProofOfPossession",
      "com.sun.identity.authentication.modules.application.Application",
      "org.forgerock.openam.authentication.modules.scripted.Scripted",
      "org.forgerock.openam.authentication.modules.social.SocialAuthOAuth2",
      "com.sun.identity.authentication.modules.hotp.HOTP",
      "org.forgerock.openam.authentication.modules.adaptive.Adaptive",
      "org.forgerock.openam.authentication.modules.accountactivecheck.AccountActiveCheck",
      "org.forgerock.openam.authentication.modules.social.SocialAuthOpenID",
      "com.sun.identity.authentication.modules.msisdn.MSISDN",
      "org.forgerock.openam.authentication.modules.fr.oath.AuthenticatorOATH",
      "com.sun.identity.authentication.modules.datastore.DataStore",
      "com.sun.identity.authentication.modules.securid.SecurID",
      "org.forgerock.openam.authentication.modules.amster.Amster",
    ],
    "defaults": {
      "accountlockout": {
        "lockoutDuration": 0,
        "lockoutDurationMultiplier": 1,
        "lockoutWarnUserCount": 0,
        "loginFailureCount": 5,
        "loginFailureDuration": 300,
        "loginFailureLockoutMode": false,
        "storeInvalidAttemptsInDataStore": true,
      },
      "core": {
        "adminAuthModule": "[Empty]",
        "orgConfig": "[Empty]",
      },
      "general": {
        "defaultAuthLevel": 0,
        "identityType": [
          "agent",
          "user",
        ],
        "locale": "en_US",
        "statelessSessionsEnabled": false,
        "twoFactorRequired": false,
        "userStatusCallbackPlugins": [],
      },
      "postauthprocess": {
        "loginFailureUrl": [],
        "loginPostProcessClass": [],
        "loginSuccessUrl": [
          "/am/console",
        ],
        "userAttributeSessionMapping": [],
        "usernameGeneratorClass": "com.sun.identity.authentication.spi.DefaultUserIDGenerator",
        "usernameGeneratorEnabled": true,
      },
      "security": {
        "addClearSiteDataHeader": true,
        "moduleBasedAuthEnabled": true,
        "sharedSecret": null,
        "zeroPageLoginAllowedWithoutReferrer": true,
        "zeroPageLoginEnabled": false,
        "zeroPageLoginReferrerWhiteList": [],
      },
      "trees": {
        "authenticationSessionsMaxDuration": 5,
        "authenticationSessionsStateManagement": "JWT",
        "authenticationSessionsWhitelist": false,
        "authenticationTreeCookieHttpOnly": true,
        "suspendedAuthenticationTimeout": 5,
      },
      "userprofile": {
        "aliasAttributeName": [],
        "defaultRole": [],
        "dynamicProfileCreation": "false",
      },
    },
    "keepPostProcessInstances": false,
    "ldapConnectionPoolDefaultSize": "1:10",
    "ldapConnectionPoolSize": [],
    "remoteAuthSecurityEnabled": false,
  },
  "meta": Any<Object>,
}
`;
